pytorch_lightning/trainer/trainer.py

```diff
@@ -1431,58 +1437,105 @@ def _call_teardown_hook(self) -> None:
         # summarize profile results
         self.profiler.describe()
 
-    def call_hook(
-        self, hook_name: str, *args: Any, pl_module: Optional["pl.LightningModule"] = None, **kwargs: Any
-    ) -> Any:
-        pl_module = self.lightning_module or pl_module
+    def _call_lightning_module_hook(
+        self,
+        hook_name: str,
+        *args: Any,
+        pl_module: Optional["pl.LightningModule"] = None,
+        **kwargs: Any,
+    ):
+        pl_module = pl_module or self.lightning_module
+
+        assert pl_module is not None, "No lightning module is available to call hooks on"
```

Comment:
Use an actual exception, asserts should be just for tests and typing enforcement

Reply:
I generally suggest `Misconfiguration` if it's actionable by the end-user (meaning, the user can do something about it and it's communicated in the message)

Otherwise one of https://docs.python.org/3/library/exceptions.html#concrete-exceptions